hail/src/main/scala/is/hail/expr/types/physical/PCanonicalStruct.scala

```diff
@@ -0,0 +1,268 @@
+              else
+                rvb.setMissing()
+              i += 1
+            }
+            fieldInserter(null, 0, rvb, inserter)
+            rvb.endStruct()
+          })
+      }
+    }
+  }
+
+  def updateKey(key: String, i: Int, sig: PType): PStruct = {
+    assert(fieldIdx.contains(key))
+
+    val newFields = Array.fill[PField](fields.length)(null)
+    for (i <- fields.indices)
+      newFields(i) = fields(i)
+    newFields(i) = PField(key, sig, i)
+    PCanonicalStruct(newFields, required)
+  }
+
+  def deleteField(key: String): PStruct = {
+    assert(fieldIdx.contains(key))
+    val index = fieldIdx(key)
+    if (fields.length == 1)
+      PCanonicalStruct(Array.empty[PField], required)
+    else {
+      val newFields = Array.fill[PField](fields.length - 1)(null)
+      for (i <- 0 until index)
+        newFields(i) = fields(i)
+      for (i <- index + 1 until fields.length)
+        newFields(i - 1) = fields(i).copy(index = i - 1)
+      PCanonicalStruct(newFields, required)
+    }
+  }
+
+  def appendKey(key: String, sig: PType): PStruct = {
+    assert(!fieldIdx.contains(key))
+    val newFields = Array.fill[PField](fields.length + 1)(null)
+    for (i <- fields.indices)
+      newFields(i) = fields(i)
+    newFields(fields.length) = PField(key, sig, fields.length)
+    PCanonicalStruct(newFields, required)
+  }
+
+
+  def rename(m: Map[String, String]): PStruct = {
+    val newFieldsBuilder = new ArrayBuilder[PField]()
+    var idx = 0
+    fields.foreach { fd =>
+      val n = fd.name
+      newFieldsBuilder += PField(m.getOrElse(n, n), fd.typ, idx)
+      idx += 1
+    }
+    PCanonicalStruct(newFieldsBuilder.result(), required)
+  }
+
+  def ++(that: PStruct): PStruct = {
+    val overlapping = fields.map(_.name).toSet.intersect(
+      that.fields.map(_.name).toSet)
+    if (overlapping.nonEmpty)
+      fatal(s"overlapping fields in struct concatenation: ${ overlapping.mkString(", ") }")
+
+    PCanonicalStruct(required, fields.map(f => (f.name, f.typ)) ++ that.fields.map(f => (f.name, f.typ)):_*)
+  }
+
+  def identBase: String = "tuple"
+
+  override def pyString(sb: StringBuilder): Unit = {
+    sb.append("struct{")
+    fields.foreachBetween({ field =>
+      sb.append(prettyIdentifier(field.name))
+      sb.append(": ")
+      field.typ.pyString(sb)
+    }) { sb.append(", ")}
+    sb.append('}')
+  }
+
+  override def _pretty(sb: StringBuilder, indent: Int, compact: Boolean) {
+    if (compact) {
+      sb.append("Struct{")
+      fields.foreachBetween(_.pretty(sb, indent, compact))(sb += ',')
+      sb += '}'
+    } else {
+      if (size == 0)
+        sb.append("Struct { }")
+      else {
+        sb.append("Struct {")
+        sb += '\n'
+        fields.foreachBetween(_.pretty(sb, indent + 4, compact))(sb.append(",\n"))
+        sb += '\n'
+        sb.append(" " * indent)
+        sb += '}'
+      }
+    }
+  }
+
+  def selectFields(names: Seq[String]): PStruct = {
+    PCanonicalStruct(required,
```

Comment:
Again, previously required wasn't being passed https://github.com/hail-is/hail/pull/7733/files#diff-bf439d99cf3638d4e2e1c9f6be564171L227